parser_library/src/parsing/parser_impl.cpp

```diff
@@ -14,453 +14,4604 @@
+                break;
+        }
+        op_name = false;
+    }
+}
+
+void parser2::process_optional_line_remark()
+{
+    if (follows<U' '>() && before_nl())
+    {
+        lex_line_remark();
+
+        adjust_lines();
+    }
+}
+
+result_t<void> parser2::process_macro_list(std::vector<semantics::concat_chain>& cc)
+{
+    assert(follows<U'('>());
+
+    consume(hl_scopes::operator_symbol);
+    if (try_consume<U')'>(hl_scopes::operator_symbol))
+        return {};
+
+    if (auto [error] = lex_macro_operand(cc.emplace_back(), true, false); error)
+        return failure;
+
+    while (try_consume<U','>(hl_scopes::operator_symbol))
+    {
+        process_optional_line_remark();
+        if (auto [error] = lex_macro_operand(cc.emplace_back(), true, false); error)
+            return failure;
+    }
+
+    if (!match<U')'>(hl_scopes::operator_symbol, diagnostic_op::error_S0011))
+        return failure;
+
+    return {};
+}
+
+result_t<void> parser2::handle_initial_space(bool reparse)
+{
+    if (!reparse && *input.next != U' ')
+    {
+        add_diagnostic(diagnostic_op::error_S0002);
+        consume_rest();
+        return failure;
+    }
+
+    // skip spaces
+    while (follows<U' '>())
+        consume();
+    adjust_lines();
+
+    return {};
+}
+
+std::pair<semantics::operand_list, range> parser2::macro_ops(bool reparse)
+{
+    const auto input_start = cur_pos_adjusted();
+    if (eof())
+        return { semantics::operand_list(), empty_range(input_start) };
+
+    if (auto [error] = handle_initial_space(reparse); error)
+        return { semantics::operand_list(), range_from(input_start) };
+
+    if (eof())
+        return { semantics::operand_list(), cur_pos_range() };
+
+    semantics::operand_list result;
+
+    auto line_start = cur_pos_adjusted();
+    auto start = line_start;
+    semantics::concat_chain cc;
+    bool pending = true;
+
+    const auto push_operand = [&]() {
+        if (!pending)
+            return;
+        if (cc.empty())
+            result.push_back(std::make_unique<semantics::empty_operand>(range_from(start)));
+        else
+            result.push_back(std::make_unique<semantics::macro_operand>(std::move(cc), range_from(start)));
+    };
+
+    // process operands
+    while (!eof())
+    {
+        switch (*input.next)
+        {
+            case U' ':
+                push_operand();
+                pending = false;
+                lex_last_remark();
+                goto end;
+
+            case U',':
+                push_operand();
+                consume(hl_scopes::operator_symbol);
+                process_optional_line_remark();
+                start = cur_pos_adjusted();
+                break;
+
+            case U'O':
+            case U'S':
+            case U'I':
+            case U'L':
+            case U'T':
+            case U'o':
+            case U's':
+            case U'i':
+            case U'l':
+            case U't':
```

Comment:
I'm not a huge fan of macros but consider creating one for these `case`s and the ones below. It might decrease the chance of missing some `case` throughout this module

Reply:
I agree that it is not ideal, but I also did not find a macro form that does not look odd/surprising at some point:
```
MACH_ATTR_CASES       // no case or colon, looks unnatural in the switch
                      // but looks natural in the macro definition
MACH_ATTR_CASES:      // looks like label in the switch
                      // macro definition still looks reasonable
case MACH_ATTR_CASES: // looks ok in the switch
                      // but the macro definition looks like a label followed by case statements.
```

Reply:
I get what you mean. I'm fine with the first two options if you decide to go with the change. I'd like to avoid macros that are difficult to read so I wouldn't consider the third one.

For the first two - I believe that a comment in the macro definition, explaining what it is, is more than fine

Reply:
As discussed, I'll keep the current version for now and revisit the issue in the future.